File: 백준/Silver/11727. 2×n 타일링 2/2×n 타일링 2.java
import java.io.*;
import java.math.*;
import java.util.*;

public class Main {

    static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

    static int N;
    static final int P = 10007;
    static int[] table;

    static void init() throws Exception {
        N = Integer.parseInt(reader.readLine());
        table = new int[N+1]; // 인덱스 1부터 쓰기
    }

    static int calc() {
        if (N == 1) return 1;

        table[1] = 1;
        table[2] = 3;
        for(int i=3; i<=N; i++) {
            table[i] = (table[i-1] + table[i-2]*2) % P;
        }
        return table[N];
    }

    public static void main(String[] args) throws Exception {
        init();
        int result = calc();
        System.out.println(result);
    }

}

File: 백준/Silver/11727. 2×n 타일링 2/2×n 타일링 2.cc
#include <iostream>

using namespace std;

int calc(int n) {
    if (n == 1) return 1;
    if (n == 2) return 3;

    int p = 10007;

    int table[1001];
    table[1] = 1;
    table[2] = 3;

    for(int i=3; i<=n; i++) {
        table[i] = (table[i-1] + table[i-2]*2) % p;
    }

    return table[n];
}

int main() {
    int n;
    cin >> n;

    int result = calc(n);
    cout << result << endl;
    return 0;
}
